contracts/core/Allo.sol

```diff
@@ -1,79 +1,284 @@
-    function createPool(PoolData memory /*_poolData*/ ) external pure returns (uint256) {
-        uint32 _poolId = 0;
+    modifier isPoolMember(uint256 _poolId) {
+        registry.isMemberOfIdentity(pools[_poolId].identityId, msg.sender);
+        _;
+    }
+
+    /// ====================================
+    /// ==== External/Public Functions =====
+    /// ====================================
 
-        // todo: return the poolId? what do we want to return here?
-        return _poolId;
+    /// @notice Fetch pool and identityMetadata
+    /// @param _poolId The id of the pool
+    /// @dev calls out to the registry to get the identity metadata
+    function getPoolInfo(uint256 _poolId) external view returns (Pool memory) {
+        return pools[_poolId];
     }
 
-    // passes _data & msg.sender through to the allocation strategy for that pool
-    // it should return a uint that represents the application
-    function applyToPool(uint256 _poolId, bytes memory _data) external payable returns (uint256) {
-        // Implement the function here
+    /// @notice Creates a new pool
+    /// @param _identityId The identityId of the pool creator in the registry
+    /// @param _allocationStrategy The address of the allocation strategy
+    /// @param _distributionStrategy The address of the distribution strategy
+    /// @param _token The address of the token that the pool is denominated in
+    /// @param _amount The amount of the token to be deposited into the pool
+    /// @param _metadata The metadata of the pool
+    function createPool(
+        bytes32 _identityId,
+        address _allocationStrategy,
+        address payable _distributionStrategy,
+        address _token,
+        uint256 _amount,
+        Metadata memory _metadata
+    ) external payable returns (uint256 poolId) {
+        if (!registry.isMemberOfIdentity(_identityId, msg.sender)) {
+            revert NO_ACCESS_TO_ROLE();
+        }
+
+        Pool memory pool = Pool({
+            identityId: _identityId,
+            allocationStrategy: IAllocationStrategy(Clone.createClone(_allocationStrategy, _nonce)),
+            distributionStrategy: IDistributionStrategy(Clone.createClone(_distributionStrategy, _nonce)),
+            metadata: _metadata
+        });
+
+        // Note: Only fund the pool on creation if the amount is greater than 0
+        if (_amount > 0) {
+            _applyFeeAndTransfer(_token, _amount, pool);
+        }
+
+        poolId = _poolIndex++;
+        pools[poolId] = pool;
+
+        emit PoolCreated(poolId, _identityId, _allocationStrategy, _distributionStrategy, _token, _amount, _metadata);
+    }
+
+    /// @notice passes _data through to the allocation strategy for that pool
+    /// @param _poolId id of the pool
+    /// @param _data encoded data unique to the allocation strategy for that pool
+    function applyToPool(uint256 _poolId, bytes memory _data) external payable returns (uint256 applicationId) {
```

Comment:
What should we return here @zobront @thelostone-mc @KurtMerbeth

Reply:
it should return the `applicationId` (received from `allocation.applyToPool`) i guess

but according to the `IAllocationStrategy` interface `bytes` are returned. We should think aboutreturning the `applicationId`

Reply:
Pairing with @thelostone-mc we came up with this:

```javascript
function applyToPool(uint256 _poolId, bytes memory _data) external payable returns (bytes memory) {
    IAllocationStrategy allocationStrategy = pools[_poolId].allocationStrategy;

    return allocationStrategy.applyToPool(_data, msg.sender);
}
```

Reply:
synced with @codenamejason. We are returning bytes cause we don't know if the applicationId would always be a unit or if an allocation strategy wanted to send a different kind of data.

Reply:
@thelostone-mc 
any ideas of what other data a strategy might return?

Reply:
paired with @KurtMerbeth. making this return uint256.

@zobront

Reply:
@thelostone-mc Agreed, returning a `uint` is fine, let's change the strategy interface to reflect that.